cpp/src/spatial/hausdorff.cu

```diff
@@ -0,0 +1,178 @@
+        }
+
+        __syncthreads(); // but we just did this, right?
+
+        if (threadIdx.x == 0)
+        {
+            // can maybe avoid this conditional by using a negate and abs trick.
+            //                 = sqrt(abs(max(-1e10, dist_sqrd_shared[0]);
+            results[block_idx] = (dist_sqrd_shared[0] < 0) ? 1e10 : sqrt(dist_sqrd_shared[0]);
+        }
+    }
+}
+
+struct hausdorff_functor
+{
+    template<typename T, typename... Args>
+    std::enable_if_t<not std::is_floating_point<T>::value, std::unique_ptr<cudf::column>>
+    operator()(Args&&...)
+    {
+        CUSPATIAL_FAIL("Non-floating point operation is not supported");
+    }
+
+    template<typename T>
+    std::enable_if_t<std::is_floating_point<T>::value, std::unique_ptr<cudf::column>>
+    operator()(cudf::column_view const& x,
+               cudf::column_view const& y,
+               cudf::column_view const& trajectory_lengths,
+               rmm::mr::device_memory_resource *mr,
+               cudaStream_t stream)
+    {
+        auto tid = cudf::experimental::type_to_id<T>();
+        auto result = cudf::make_fixed_width_column(cudf::data_type{ tid },
+                                                    trajectory_lengths.size() * trajectory_lengths.size(),
+                                                    cudf::mask_state::UNALLOCATED,
+                                                    stream,
+                                                    mr);
+
+        auto d_x = cudf::column_device_view::create(x);
+        auto d_y = cudf::column_device_view::create(y);
+        auto d_trajectory_lengths = cudf::column_device_view::create(trajectory_lengths);
+        auto d_trajectory_offsets = rmm::device_vector<cudf::size_type>(trajectory_lengths.size());
+
+        thrust::exclusive_scan(rmm::exec_policy(stream)->on(stream),
+                                                d_trajectory_lengths->begin<cudf::size_type>(),
+                                                d_trajectory_lengths->end<cudf::size_type>(),
+                                                d_trajectory_offsets.begin());
+
+        return result;
+    }
+};
+
+} // namespace anonymous
+
+namespace cuspatial {
+
+std::unique_ptr<cudf::column>
+directed_hausdorff_distance(cudf::column_view const& x,
+                            cudf::column_view const& y,
+                            cudf::column_view const& trajector_lengths,
+                            rmm::mr::device_memory_resource *mr)
+{
+    CUSPATIAL_EXPECTS(x.type() == y.type(), "inputs `x` and `y` must have same type.");
+    CUSPATIAL_EXPECTS(x.size() == y.size(), "inputs `x` and `y` must have same length.");
+
+    CUSPATIAL_EXPECTS(not x.has_nulls() and
+                      not y.has_nulls() and
+                      not trajector_lengths.has_nulls(),
+                      "inputs must not have nulls.");
+
+    CUSPATIAL_EXPECTS(x.size() >= trajector_lengths.size(),
```

Comment:
I suppose this is an easy check that can fail early, but I think it should still check if the sum of all elements in `trajectory_length` is <= `x.size()` (or maybe it should be exactly equal)? Since you're already computing `exclusive_scan` of `trajectory_length` inside `hausdorff_functor()` anyway, maybe you could change it to an `inclusive_scan` so you can grab the sum as `d_trajectory_offsets.back()`. You can use `d_trajectory_offsets.begin() + 1` for the output iterator and manually set the first element to zero so it still looks like an `inclusive_scan`.

Reply:
That's a device to host copy which requires synchronization.